File: lib/tradingReports/tradingReportPeriods.ts
import type { TradingReportKind, TradingReportPeriodKey } from "./tradingReportTypes"

export const TRADING_REPORT_PERIOD_OPTIONS: ReadonlyArray<{
  key: TradingReportPeriodKey
  label: string
  kind: TradingReportKind
}> = [
  { key: "weekly_this", label: "This Week", kind: "weekly" },
  { key: "weekly_last", label: "Last Week", kind: "weekly" },
  { key: "monthly_this", label: "This Month", kind: "monthly" },
  { key: "monthly_last", label: "Last Month", kind: "monthly" },
]

export function startOfDay(date: Date): Date {
  const next = new Date(date)
  next.setHours(0, 0, 0, 0)
  return next
}

export function endOfDay(date: Date): Date {
  const next = new Date(date)
  next.setHours(23, 59, 59, 999)
  return next
}

/** Monday 00:00 of the calendar week containing `date`. */
export function getMondayOfWeek(date: Date): Date {
  const monday = startOfDay(date)
  const day = monday.getDay()
  const diff = day === 0 ? -6 : 1 - day
  monday.setDate(monday.getDate() + diff)
  return monday
}

export function getSundayOfWeek(monday: Date): Date {
  const sunday = new Date(monday)
  sunday.setDate(monday.getDate() + 6)
  return endOfDay(sunday)
}

export type TradingReportPeriodBounds = {
  key: TradingReportPeriodKey
  kind: TradingReportKind
  start: Date
  end: Date
}

export function getTradingReportPeriodBounds(
  key: TradingReportPeriodKey,
  now = new Date()
): TradingReportPeriodBounds {
  const today = startOfDay(now)

  if (key === "weekly_this" || key === "weekly_last") {
    let monday = getMondayOfWeek(today)
    if (key === "weekly_last") {
      monday = new Date(monday)
      monday.setDate(monday.getDate() - 7)
    }

    const sunday = getSundayOfWeek(monday)
    const end = key === "weekly_this" ? endOfDay(now) : sunday

    return { key, kind: "weekly", start: monday, end }
  }

  if (key === "monthly_this") {
    const start = startOfDay(new Date(now.getFullYear(), now.getMonth(), 1))
    return { key, kind: "monthly", start, end: endOfDay(now) }
  }

  const start = startOfDay(new Date(now.getFullYear(), now.getMonth() - 1, 1))
  const end = endOfDay(new Date(now.getFullYear(), now.getMonth(), 0))
  return { key, kind: "monthly", start, end }
}

export function formatTradingReportDateRange(
  start: Date,
  end: Date,
  kind: TradingReportKind
): string {
  const sameDay = start.toDateString() === end.toDateString()
  if (sameDay) {
    return start.toLocaleDateString(undefined, {
      month: "short",
      day: "numeric",
      year: "numeric",
    })
  }

  if (kind === "monthly" && start.getDate() === 1) {
    const monthEnd = endOfDay(
      new Date(start.getFullYear(), start.getMonth() + 1, 0)
    )
    if (end.toDateString() === monthEnd.toDateString()) {
      return start.toLocaleDateString(undefined, {
        month: "long",
        year: "numeric",
      })
    }
  }

  const startYear = start.getFullYear()
  const endYear = end.getFullYear()
  const startLabel = start.toLocaleDateString(undefined, {
    month: "short",
    day: "numeric",
    ...(startYear !== endYear ? { year: "numeric" } : {}),
  })
  const endLabel = end.toLocaleDateString(undefined, {
    month: "short",
    day: "numeric",
    year: "numeric",
  })
  return `${startLabel} – ${endLabel}`
}

export function tradingReportPeriodTitle(kind: TradingReportKind): string {
  return kind === "weekly" ? "Weekly Trading Report" : "Monthly Trading Report"
}

/** Stable id for notification dedup — weekly = Monday ISO date, monthly = YYYY-MM */
export function tradingReportPeriodId(
  key: TradingReportPeriodKey,
  now = new Date()
): string {
  const { start, kind } = getTradingReportPeriodBounds(key, now)
  if (kind === "weekly") {
    return `week:${start.toISOString().slice(0, 10)}`
  }
  const month = String(start.getMonth() + 1).padStart(2, "0")
  return `month:${start.getFullYear()}-${month}`
}

export function isWeeklyReportReleaseDay(now = new Date()): boolean {
  return now.getDay() === 1
}

export function isMonthlyReportReleaseDay(now = new Date()): boolean {
  return now.getDate() === 1
}

File: native-ios/TradeTraxs/TradeTraxs/Domain/Services/TradingReportPeriods.swift
import Foundation

/// Web `lib/tradingReports/tradingReportPeriods.ts` — calendar week (Mon–Sun) and month windows.
nonisolated enum TradingReportPeriods {
    struct Bounds: Hashable, Sendable {
        var key: TradingReportPeriodKey
        var kind: TradingReportKind
        var start: Date
        var end: Date
    }

    static func bounds(
        for key: TradingReportPeriodKey,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> Bounds {
        let today = calendar.startOfDay(for: now)

        switch key {
        case .weeklyThis, .weeklyLast:
            var monday = mondayOfWeek(containing: today, calendar: calendar)
            if key == .weeklyLast {
                monday = calendar.date(byAdding: .day, value: -7, to: monday) ?? monday
            }
            let sunday = endOfDay(
                calendar.date(byAdding: .day, value: 6, to: monday) ?? monday,
                calendar: calendar
            )
            let end = key == .weeklyThis ? endOfDay(now, calendar: calendar) : sunday
            return Bounds(key: key, kind: .weekly, start: monday, end: end)

        case .monthlyThis:
            let comps = calendar.dateComponents([.year, .month], from: today)
            let start = calendar.date(from: DateComponents(
                year: comps.year,
                month: comps.month,
                day: 1
            )) ?? today
            return Bounds(
                key: key,
                kind: .monthly,
                start: calendar.startOfDay(for: start),
                end: endOfDay(now, calendar: calendar)
            )

        case .monthlyLast:
            let comps = calendar.dateComponents([.year, .month], from: today)
            let thisMonthStart = calendar.date(from: DateComponents(
                year: comps.year,
                month: comps.month,
                day: 1
            )) ?? today
            let start = calendar.date(byAdding: .month, value: -1, to: thisMonthStart) ?? thisMonthStart
            let endDay = calendar.date(byAdding: .day, value: -1, to: thisMonthStart) ?? start
            return Bounds(
                key: key,
                kind: .monthly,
                start: calendar.startOfDay(for: start),
                end: endOfDay(endDay, calendar: calendar)
            )
        }
    }

    static func dateRangeLabel(start: Date, end: Date, kind: TradingReportKind) -> String {
        let calendar = Calendar.current
        if calendar.isDate(start, inSameDayAs: end) {
            return Self.medium.string(from: start)
        }

        if kind == .monthly, calendar.component(.day, from: start) == 1 {
            let monthEnd = calendar.date(
                byAdding: DateComponents(month: 1, day: -1),
                to: calendar.startOfDay(for: start)
            )
            if let monthEnd, calendar.isDate(end, inSameDayAs: monthEnd) {
                return Self.monthYear.string(from: start)
            }
        }

        let startYear = calendar.component(.year, from: start)
        let endYear = calendar.component(.year, from: end)
        let startLabel: String
        if startYear != endYear {
            startLabel = Self.medium.string(from: start)
        } else {
            startLabel = Self.monthDay.string(from: start)
        }
        let endLabel = Self.medium.string(from: end)
        return "\(startLabel) – \(endLabel)"
    }

    /// Calendar year window — Jan 1 through Dec 31, or through today for the current year.
    static func yearBounds(
        year: Int,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> Bounds {
        let start = calendar.date(from: DateComponents(year: year, month: 1, day: 1)) ?? now
        let startDay = calendar.startOfDay(for: start)
        let currentYear = calendar.component(.year, from: now)
        let end: Date
        if year == currentYear {
            end = endOfDay(now, calendar: calendar)
        } else if year < currentYear {
            let nextYearStart = calendar.date(from: DateComponents(year: year + 1, month: 1, day: 1)) ?? start
            let lastDay = calendar.date(byAdding: .day, value: -1, to: nextYearStart) ?? start
            end = endOfDay(lastDay, calendar: calendar)
        } else {
            end = endOfDay(start, calendar: calendar)
        }
        return Bounds(key: .monthlyThis, kind: .monthly, start: startDay, end: end)
    }

    /// Single calendar month — full month for past years; current month ends today.
    static func monthBounds(
        year: Int,
        month: Int,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> Bounds {
        let start = calendar.date(from: DateComponents(year: year, month: month, day: 1)) ?? now
        let startDay = calendar.startOfDay(for: start)
        let currentYear = calendar.component(.year, from: now)
        let currentMonth = calendar.component(.month, from: now)
        let end: Date
        if year == currentYear, month == currentMonth {
            end = endOfDay(now, calendar: calendar)
        } else {
            let nextMonthStart = calendar.date(byAdding: .month, value: 1, to: startDay) ?? startDay
            let lastDay = calendar.date(byAdding: .day, value: -1, to: nextMonthStart) ?? startDay
            end = endOfDay(lastDay, calendar: calendar)
        }
        return Bounds(key: .monthlyThis, kind: .monthly, start: startDay, end: end)
    }

    /// Years that contain at least one non-backtest trade (newest first).
    static func availableYears(
        from trades: [Trade],
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> [Int] {
        var years = Set<Int>()
        for trade in trades where trade.mode != .backtest {
            years.insert(calendar.component(.year, from: trade.entryAt))
        }
        return years.sorted(by: >)
    }

    static func yearDateRangeLabel(
        year: Int,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> String {
        let bounds = yearBounds(year: year, now: now, calendar: calendar)
        let currentYear = calendar.component(.year, from: now)
        if year == currentYear {
            let formatter = DateFormatter()
            formatter.locale = .current
            formatter.setLocalizedDateFormatFromTemplate("MMMMyyyy")
            let startLabel = formatter.string(from: bounds.start)
            let endLabel = Self.medium.string(from: bounds.end)
            return "\(startLabel) – \(endLabel)"
        }
        let formatter = DateFormatter()
        formatter.locale = .current
        formatter.setLocalizedDateFormatFromTemplate("yyyy")
        return formatter.string(from: bounds.start)
    }

    /// Web `tradingReportPeriodId` for notify dedupe.
    static func periodID(for key: TradingReportPeriodKey, now: Date = Date()) -> String {
        let bounds = bounds(for: key, now: now)
        switch bounds.kind {
        case .weekly:
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withFullDate]
            return "week:\(formatter.string(from: bounds.start))"
        case .monthly:
            let comps = Calendar.current.dateComponents([.year, .month], from: bounds.start)
            let month = String(format: "%02d", comps.month ?? 1)
            return "month:\(comps.year ?? 0)-\(month)"
        }
    }

    // MARK: - Helpers

    private static func mondayOfWeek(containing date: Date, calendar: Calendar) -> Date {
        let start = calendar.startOfDay(for: date)
        let weekday = calendar.component(.weekday, from: start) // 1 = Sunday
        let diff = weekday == 1 ? -6 : 1 - weekday
        return calendar.date(byAdding: .day, value: diff, to: start) ?? start
    }

    private static func endOfDay(_ date: Date, calendar: Calendar) -> Date {
        let start = calendar.startOfDay(for: date)
        guard let next = calendar.date(byAdding: .day, value: 1, to: start) else { return date }
        return next.addingTimeInterval(-0.001)
    }

    private static let medium: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = .current
        formatter.setLocalizedDateFormatFromTemplate("MMMdyyyy")
        return formatter
    }()

    private static let monthDay: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = .current
        formatter.setLocalizedDateFormatFromTemplate("MMMd")
        return formatter
    }()

    private static let monthYear: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = .current
        formatter.setLocalizedDateFormatFromTemplate("MMMMyyyy")
        return formatter
    }()
}
